# HG changeset patch
# User John W. Eaton <jwe@octave.org>
# Date 1219262234 14400
#      Wed Aug 20 15:57:14 2008 -0400
# Node ID 24701aa75ecbe5f64da30b756890888a858ef029
# Parent  faf0abc5fd515423c7b95083b03260bdfcc6f777
scope fixes for anonymous and inline functions that appear inside subfunctions

diff --git a/src/ov-fcn-inline.cc b/src/ov-fcn-inline.cc
--- a/src/ov-fcn-inline.cc
+++ b/src/ov-fcn-inline.cc
@@ -86,17 +86,30 @@ octave_fcn_inline::octave_fcn_inline (co
 
       if (fh)
 	{
 	  fcn = fh->fcn_val ();
 
 	  octave_user_function *uf = fcn.user_function_value ();
 
 	  if (uf)
-	    uf->stash_parent_fcn_scope (octave_call_stack::current_scope ());
+	    {
+	      octave_function *curr_fcn = octave_call_stack::current ();
+
+	      if (curr_fcn)
+		{
+		  symbol_table::scope_id parent_scope
+		    = curr_fcn->parent_fcn_scope ();
+
+		  if (parent_scope < 0)
+		    parent_scope = curr_fcn->scope ();
+	
+		  uf->stash_parent_fcn_scope (parent_scope);
+		}
+	    }
 	}
     }
 
   if (fcn.is_undefined ())
     error ("inline: unable to define function");
 }
 
 // This function is supplied to allow a Matlab style class structure
diff --git a/src/pt-fcn-handle.cc b/src/pt-fcn-handle.cc
--- a/src/pt-fcn-handle.cc
+++ b/src/pt-fcn-handle.cc
@@ -109,17 +109,23 @@ tree_anon_fcn_handle::rvalue (void)
 				ret_list ? ret_list->dup (new_scope, 0) : 0,
 				cmd_list ? cmd_list->dup (new_scope, 0) : 0);
 
   octave_function *curr_fcn = octave_call_stack::current ();
 
   if (curr_fcn)
     {
       uf->stash_parent_fcn_name (curr_fcn->name ());
-      uf->stash_parent_fcn_scope (curr_fcn->scope ());
+
+      symbol_table::scope_id parent_scope = curr_fcn->parent_fcn_scope ();
+
+      if (parent_scope < 0)
+	parent_scope = curr_fcn->scope ();
+	
+      uf->stash_parent_fcn_scope (parent_scope);
     }
 
   uf->mark_as_inline_function ();
 
   octave_value ov_fcn (uf);
 
   octave_value fh (new octave_fcn_handle (ov_fcn, "@<anonymous>"));
 
